draft-ietf-mls-protocol.md

```diff
@@ -1008,14 +1112,178 @@ struct {
+
+Handshake and application messages use a common framing structure.
+This framing provides encryption to assure confidentiality within the
+group, as well as signing to authenticate the sender within the group.
+
+The two main structures involved are MLSPlaintext and MLSCiphertext.
+MLSCiphertext represents a signed and encrypted message, with
+protections for both the content of the message and related
+metadata.  MLSPlaintext represents a message that is only signed,
+and not encrypted.  Applications SHOULD use MLSCiphertext to encode
+both application and handshake messages, but MAY transmit handshake
+messages encoded as MLSPlaintext objects in cases where it is
+necessary for the delivery service to examine such messages.
+
+~~~~~
+enum {
+    invalid(0),
+    handshake(1),
+    application(2),
+    (255)
+} ContentType;
+
+struct {
+    opaque group_id<0..255>;
+    uint32 epoch;
+    uint32 sender;
+    ContentType content_type;
+
+    select (MLSPlaintext.type) {
+        case handshake:
+            GroupOperation operation;
+
+        case application:
+            opaque application_data<0..2^32-1>;
+    }
+
+    opaque signature<0..2^16-1>;
+} MLSPlaintext;
+
+struct {
+    opaque group_id<0..255>;
+    uint32 epoch;
+    ContentType content_type;
+    opaque masked_sender_data[8];
+    opaque ciphertext<0..2^32-1>;
+} MLSCiphertext;
+~~~~~
+
+The remainder of this section describe how to compute the signature of
+an MLSPlaintext object and how to convert it to an MLSCiphertext object.
+The overall process is as follows:
+
+* Compute the metadata by concatenating the following information:
+  * Group ID
+  * Epoch
+  * Sender index
+  * Key generation
+
+* Sign the protected content and metadata
+
+* Encrypt the content using a content encryption key identified by
+  the metadata
+
+* Mask the sender_data used to identify the content encryption key
+  using a per-epoch key
+
+The group identifier and epoch fields are copied from the MLSPlaintext object directly.
+The content encryption process populates the ciphertext field of the
+MLSCiphertext object.  The metadata encryption step populates the
+masked_sender_data field.
+
+Decryption follows the same step in reverse: Unmasking the
+metadata, then decrypting the content, then verifying the content
+signature.
+
+
+## Content Signing and Encryption
+
+The signature field in an MLSPlaintext object is computed using the
+signing private key corresponding to the credential at the leaf in
+the tree indicated by the sender field.  The signature covers the
+MLSPlaintext object, with the signature field truncated.
+
+The ciphertext field of the MLSCiphertext object is produced by
+supplying the inputs described below to the AEAD function specified
+by the ciphersuite in use.  The plaintext input contains content and
+signature of the MLSPlaintext, plus optional padding.  These values
+are encoded in the following form:
+
+~~~~~
+struct {
+    opaque content[MLSPlaintext.length];
```

Comment:
It occurs to me that there's not actually a `length` field in `MLSPlaintext`.  Let's just use `length_of_content`, in parallel to `length_of_padding`.